### PR TITLE
Monitor repairing threshhold

### PR DESCRIPTION
**Pull request by JeffreyDevloo**
#127 - object repair test
Added service helper method
Removed unused code in logger
Added custom value logging
Added a primitive caching system
Fixed KeyError issue: thrown keyerrors were caught by the run method and returned the wrong output

### CODE REVIEW

File: ovs/extensions/healthcheck/alba/alba_health_check.py [hunk 1]
```diff
@@ -400,6 +402,145 @@ def check_alba(logger):
...
+        test_name = 'disk-safety'
+        result = {
+                "repair_percentage": None,
+                "lost_disks": None
+            }
+        for service in ServiceHelper.get_services():
+            if service.type.name == ServiceType.SERVICE_TYPES.ALBA_MGR:
+                abms.append(service.name)
+
+        abms = list(set(abms))
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/alba/alba_health_check.py]: why would you put this in a set first and then in a list?

File: ovs/extensions/healthcheck/alba/alba_health_check.py [hunk 2]
```diff
@@ -400,6 +402,145 @@ def check_alba(logger):
...
+            if service_name not in abms:
+                continue
+
+            config = "arakoon://config/ovs/arakoon/{0}/config?ini=%2Fopt%2FOpenvStorage%2Fconfig%2Farakoon_cacc.ini".format(service_name)
+
+            # Fetch alba info
+            try:
+                namespaces = AlbaCLI.run('show-namespaces', config=config, to_json=True)[1]
+            except Exception as ex:
+                raise RuntimeError("Could not execute 'alba show-namespaces'. Got {0}".format(ex.message))
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/alba/alba_health_check.py]: What if the abm arakoons are down? The healthcheck just stops running? You should provide a `[EXCEPTION]` message to the customer.

File: ovs/extensions/healthcheck/alba/alba_health_check.py [hunk 3]
```diff
@@ -400,6 +402,145 @@ def check_alba(logger):
...
+
+            try:
+                presets = AlbaCLI.run('list-presets', config=config, to_json=True)
+            except Exception as ex:
+                raise RuntimeError("Could not execute 'list-presets'. Got {0}".format(ex.message))
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/alba/alba_health_check.py]: What if the abm arakoons are down? The healthcheck just stops running? You should provide a `[EXCEPTION]` message to the customer.

File: ovs/extensions/healthcheck/alba/alba_health_check.py [hunk 4]
```diff
@@ -400,6 +402,145 @@ def check_alba(logger):
...
+                total_objects = result["fields"]["total_objects"]
+                # Ignore fully healthy ones
+                if result["tags"]["disk_lost"] != 0:
+                    current_disks_lost = result["tags"]["disk_lost"]
+                    object_to_be_repaired = result["fields"]["objects"]
+            repair_percentage = object_to_be_repaired / total_objects
+            if current_disks_lost == 0:
+                logger.success("Found no losts disks. All data is safe.")
+            else:
+                logger.failure("Currently found {0} disks that are lost. Alba might or might not repair these disks.".format(current_disks_lost))
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/alba/alba_health_check.py]: Alba might or might not repair these disks? Alba will repair you data ;)!

File: ovs/extensions/healthcheck/helpers/cache.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    @staticmethod
+    def get(key=prefix):
+        try:
+            return CacheHelper._from_json(CacheHelper.client.get(key))
+        except KeyNotFoundException:
+            return None
+
+    @staticmethod
+    def _to_json(value):
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/helpers/cache.py]: to_json method but you return the plain value if you receive a type error. This method will (in most cases by invalid) if a person is developing. Here you are permitted to raise a `RuntimeError` or reraising the `TypeError`. We want to know if a give json is invalid.

File: ovs/extensions/healthcheck/helpers/cache.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        try:
+            json_object = json.loads(str(value))
+        except ValueError:
+            return json.dumps(value)
+        except TypeError:
+            return value
+        return json_object
+
+    @staticmethod
+    def _from_json(value):
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/helpers/cache.py]: see `to_json` comment. Applies here too.

File: ovs/lib/healthcheck.py [hunk 7]
```diff
@@ -375,40 +375,38 @@ def run_method(module_name=None, method_name=None, *args):
...
+        return HealthCheckController.print_methods()
 
     if __name__ == '__main__':
         import sys
         from ovs.lib.healthcheck import HealthCheckController
         arguments = sys.argv
         # Remove filename
         del arguments[0]
-        #arguments = ('alba', 'proxy-test')
+        # arguments = ('alba', 'disk-safety')
```
Comment (kinvaris) [inline ovs/lib/healthcheck.py]: remove this line

File: ovs/log/healthcheck_logHandler.py [hunk 8]
```diff
@@ -78,7 +80,7 @@ def __init__(self, print_progress=True):
 
         self._logger = LogHandler.get("healthcheck")
 
-    def _log(self, msg, test_name, error_message=None):
+    def _log(self, msg, test_name, error_message=None, value=None):
```
Comment (kinvaris) [inline ovs/log/healthcheck_logHandler.py]: maybe change to custom_value?

Comment (kinvaris) [inline ovs/log/healthcheck_logHandler.py]: also don't forget the docstring!

File: ovs/extensions/healthcheck/alba/alba_health_check.py (hunk 1 above)
Comment (JeffreyDevloo) [inline ovs/extensions/healthcheck/alba/alba_health_check.py]: Was to only have unique values.
However the check can happen when appending to abms in the first place.

File: ovs/extensions/healthcheck/helpers/cache.py [hunk 9]
```diff
@@ -31,45 +31,69 @@ def set(info):
...
         except TypeError:
-            return value
-        return json_object
+            # Not a dict
+            return data
+        return parsed_data
 
     @staticmethod
-    def _from_json(value):
+    def _test_type_handling():
```
Comment (kinvaris) [inline ovs/extensions/healthcheck/helpers/cache.py]: Isn't this for unittesting?

Comment (JeffreyDevloo) [inline ovs/extensions/healthcheck/helpers/cache.py]: It is a minor smoke test.

**Pull request merged**